### PR TITLE
fmt parser attempt # ∞

### PR DESCRIPTION
**Pull request by cmyr**
This is a skeletal version of a parser and matcher for format strings. A bunch of cleanup to do, but I think this is a reasonable overall approach.

Will try and do some cleanup and get this integrated now.

### CODE REVIEW

File: pars-derive/src/lib.rs [hunk 1]
```diff
@@ -174,7 +178,50 @@ fn generate_fmt_block(
...
+    }
+
+    let field_names = cont.data.all_fields().filter_map(|f| match &f.member {
+        syn::Member::Named(ident) => Some(ident.to_string()),
+        _ => None,
+    }).collect::<Vec<_>>();
+    let fmt_string = get_attr_string(attrs).map_err(|e| vec![e])?;
+    let num_fields = cont.data.num_fields();
+
+    let _ = ::pars_fmt::FmtMatcher::new(&fmt_string, field_names.as_slice()).unwrap();
```
Comment (gosulliv) [inline pars-derive/src/lib.rs:197]: What's this about?

File: pars-fmt/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+}
+
+#[derive(Debug, Clone, PartialEq)]
+enum Token {
+    Field(String),
+    Separator(String),
+}
+
+struct Parser<'a> {
+    inp: &'a str,
```
Comment (gosulliv) [inline pars-fmt/src/lib.rs]: This name makes sense to someone who's used to thinking of pointers from C, but maybe we could call it `source` or somesuch?

**Review by gosulliv — APPROVED** [review]
I think this would be easier to read with `chars()` or some other iter, that being said, it looks like it works, let's get it in for the talk!

**Comment by gosulliv** [comment]
also actuall implementing the iterator in a way that's correct and looks good would be hard.

File: pars-derive/src/lib.rs (hunk 1 above)
Comment (cmyr) [reply pars-derive/src/lib.rs:197]: just validation of the fmt string at compile time, we do the same thing with the regex.

**Comment by cmyr** [comment]
okay, gave it another pass. The bit with not using chars() is just me getting fancy. 🤷‍♂️

File: pars-derive/src/lib.rs (hunk 1 above)
Comment (gosulliv) [reply pars-derive/src/lib.rs:197]: ah, got it. Thanks for the comment.

**Comment by gosulliv** [comment]
Yeah, and me complaining about not using `chars()` is just my own dislike of index arithmetic because it's easy to get wrong :)

:shipit:

**Pull request merged**